### PR TITLE
Proposed fix for Error when mapping nested expressions #2201

### LINKED ISSUE #2201: Error when mapping expressions with array members
When I try to map nested expressions, AutoMapper throws an exception.

### Expected behavior
Correctly mapped expression

### Actual behavior
Exception

### Steps to reproduce

I created a gist that reproduces the error: 
https://gist.github.com/smax48/005e3121206b69dcb666978fd418c7c0

## Update
It seems that the problem is not in nested expressions, but rather in the used collection type. When it is IEnumerable<T> or IList<T>, it works OK. When it's array - it doesn't ...

### PR DESCRIPTION
**Pull request by BlaiseD**
Fixes #2201 -resolves type mappings for underlying array elements.

### CODE REVIEW

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs [hunk 1]
```diff
@@ -176,7 +176,18 @@ public static List<ParameterExpression> GetDestinationParameterExpressions(this
         public static Dictionary<Type, Type> AddTypeMapping<TSource, TDest>(this Dictionary<Type, Type> typeMappings)
             => typeMappings == null
                 ? throw new ArgumentException(Resource.typeMappingsDictionaryIsNull)
-                : typeMappings.AddTypeMapping(typeof(TSource), typeof(TDest));
+                : typeMappings.AddTypeMapping(typeof(TSource), typeof(TDest));
+
+        private static void AddUnderlyimgArrayElement(this Dictionary<Type, Type> typeMappings, Type sourceType, Type destType)
```
Comment (smax48) [inline src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: Typo in the function name - should be AddUnderlyingArrayElement (same for the next one: AddUnderlyingGenericTypes)

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs [hunk 2]
```diff
@@ -176,7 +176,18 @@ public static List<ParameterExpression> GetDestinationParameterExpressions(this
         public static Dictionary<Type, Type> AddTypeMapping<TSource, TDest>(this Dictionary<Type, Type> typeMappings)
             => typeMappings == null
                 ? throw new ArgumentException(Resource.typeMappingsDictionaryIsNull)
-                : typeMappings.AddTypeMapping(typeof(TSource), typeof(TDest));
+                : typeMappings.AddTypeMapping(typeof(TSource), typeof(TDest));
+
+        private static void AddUnderlyimgArrayElement(this Dictionary<Type, Type> typeMappings, Type sourceType, Type destType)
+        {
+            if (sourceType.IsArray && destType.IsArray)
```
Comment (lbargaoanu) [inline src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: Maybe consider all collections and use [ElementTypeHelper](https://github.com/AutoMapper/AutoMapper/blob/524f587f51d43d3cb17aacfaeffb41b70f93c39c/src/AutoMapper/Mappers/Internal/ElementTypeHelper.cs)?

**Comment by BlaiseD** [comment]
Done - thanks for the heads-up.

**Comment by lbargaoanu** [comment]
https://github.com/blog/1506-closing-issues-via-pull-requests

**Pull request merged**